Expand kubeclient interface to avoid leaking goroutines in Upgrade/Install RunWithContext when context is cancelled

saisrikark opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

closes #13126

**What this PR does / why we need it**:
When using RunWithContext for Upgrade/Install with a Wait/WaitWithJobs/Atomic set it is possible for goroutines to leak/remain dangling when the context is cancelled prematurely.

This is due to these methods not waiting for the helm kubeclient's Wait and WaitWithJobs to complete.

With this PR I am introducing new methods in the kubeclient interface.
WaitWithContext and WaitWithJobsWithContext
This way a context is propagated and goroutines are exited solving dangling resources.

**Special notes for your reviewer**:
Snippet to try this out

```
package main

import (
	"context"
	"log"
	"time"

	"helm.sh/helm/v3/pkg/action"
)


func main() {

	installAction := action.NewInstall(&action.Configuration{
		// fill this up
		// ensure to use a logger to display the log in the dangling go routine
	})

	installAction.Wait = true
	installAction.Timeout = time.Hour

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()

	// fill up the chart and values
	go func() {
		_, err := installAction.RunWithContext(ctx, "chart", "values")
		if err != nil {
			log.Printf("encountered an error %s", err.Error())
		}
	}()

	waitChan := make(chan struct{})
	<-waitChan
}
```

saisrikark committed:
Expand kubeclient interface to avoid leaking goroutines in Upgrade/Install RunWithContext when context is cancelled

Signed-off-by: saisrikark <saisrikark@gmail.com>

robertsirc commented:
I am still checking to see if this is a breaking change since the contract has changed. Is there an opportunity of having testing for this change?

saisrikark commented:
Testcases are a good idea.
Let me try it out.

gjenkins8 reviewed (commented):
Wouldn't we want implement e.g. `Wait` in terms of `WaitWithContext` ?

```
func (c *Client) Wait(resources ResourceList, timeout time.Duration) error {
   return WaitWithContext(context.Background(), resources, timeout)
}
```

(and others of course)

gjenkins8 reviewed (commented):
Can we please add a comment to the replaced ie. non-context methods, saying to the effect that they are deprecated and due for removal in Helm v4

saisrikark committed:
modifying tests and removing leaking code

Signed-off-by: saisrikark <saisrikark@gmail.com>

saisrikark committed:
Merge pull request #2 from saisrikark/context-13126-helm

modifying tests and removing leaking code

saisrikark commented:
Added testcases and addressed comments.
handleContext looks to be used to handle sigterm and sigint.

I wonder what is the expected behaviour when we see these kind of signals.

I feel we should store information about the release (which elucidates the failure reason) and stop other operations.
Will update this PR once I have a better understanding.

saisrikark committed:
precheck to fail for a cancelled context

Signed-off-by: saisrikark <saisrikark@gmail.com>

saisrikark committed:
fixing lint issues

Signed-off-by: saisrikark <saisrikark@gmail.com>
